testing-guidelines/ci.md

```diff
@@ -0,0 +1,431 @@
+
+- Which workflow file on which branch actually runs.
+- What information about the repo or triggering event does that workflow have access to.
+
+If a workflow "runs in the context of the default branch" then it will use the workflow that exists on trhe default branch, regardless of whether or not the event that originall triggered it was on another branch. If the workflow "runs in the context of the pull request branch" then it will pull the workflow file from the pull request branch.
+
+The information available inside a workflow after it has started (usually available via the [`github` context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context)). For pull requests, this will include things like the pull request number, and the ref and HEAD SHA of the pull request branch. For workflows running in the context of the default branch, this may not contain much information, but all references to the branch and sha will mainly just be references to main.
+
+</details>
+
+#### New solution, new problems
+
+For the reasons described above, we chose to use `workflow_run` _heavily_ for the gradio repo. However `workflow_run` presents its own challenges:
+
+- This event runs in the context of main, it doesn't offer any of the conveniences that `push` and `pull_request` events give you, it knows very very little about the workflow run even that triggered it. It _does not_ inherit the teriggering workflow's context. This is a huge problem.
+- This workflow kind of runs in the void. It is run in the context of the default branch and so maintains references to that branch, however, it isn't really 'attached' to a commit or ref in any meaningful way and the status of the run (the 'check') is not added to any commits anywhere.
+
+Both of these problems were eventually solve by using the GitHub API in combination with the information we get from the workflow event's context. Getting the commit reference of the pull request that triggered the workflow is the main challenge, when we have that, creating statuses on commits is trivial. 
+
+##### What branch am i even in?
+
+The 'context' a workflow runs in is the branch that actions/checkout will checkout by default. In this case that is `main`. We don't want `main`.
+
+Figuring out what branch or pull request triggered a workflow run is surprisingly difficult depending on the event that you allow to trigger it. The data you have access to in a ` workflow_run` event is pretty limited. It is okay for pull requests, you get the PR number, but for pushes and other events it can be challenging. We trigger `workflow_runs` from the following events:
+
+- `pull_request` - this is fine, we get the PR number.
+- `push` - not fine, but we get the commit SHA which will do.
+- `issue_comment` - couldn't be less fine, we only get the title.
+
+*The reason we use the `issue_comment` event is because pull request are actually a special type of issue when it comes to the title, body, and replies.*
```

Comment:
(GitHub likes this pattern. Similarly, organizations on GitHub are a special kind of user.)